cpp/include/raft/spatial/knn/detail/ann_kmeans_balanced.cuh

```diff
@@ -759,30 +808,30 @@ void build_optimized_kmeans(const handle_t& handle,
   RAFT_EXPECTS(n_clusters_done == n_clusters, "Didn't process all clusters.");
 
   rmm::device_uvector<uint32_t> cluster_sizes(n_clusters, stream, device_memory);
-  rmm::device_uvector<uint32_t> labels(n_rows_train, stream, device_memory);
-
-  // fit clusters using the trainset
-  for (int iter = 0; iter < 2; iter++) {
-    predict(handle,
-            cluster_centers,
-            n_clusters,
-            dim,
-            trainset.data(),
-            n_rows_train,
-            labels.data(),
-            metric,
-            stream,
-            device_memory);
-    calc_centers_and_sizes(cluster_centers,
-                           cluster_sizes.data(),
-                           n_clusters,
-                           dim,
-                           trainset.data(),
-                           n_rows_train,
-                           labels.data(),
-                           true,
-                           stream);
-  }
+  rmm::device_uvector<uint32_t> labels(n_rows, stream, device_memory);
+
+  // Fine-tuning kmeans for all clusters
+  //
+  // (*) Since the likely cluster centroids have been calculated
+  // hierarchically already, the number of iteration for fine-tuning
+  // kmeans for whole clusters should be reduced. However, there
+  // is a possibility that the clusters could be unbalanced here,
+  // in which case the actual number of iterations would be increased.
+  //
+  balancing_em_iters(handle,
+                     std::max<uint32_t>(n_iters / 10, 2),
```

Comment:
Is `n_iters / 10` the value that was used in IVF-PQ?

Reply:
Yes, and I left the comment from cuann repo as-is. I figure, this fine-tuning step isn't really specific to the ivf-pq use-case, so I've backported the changes here.